modules/imgcodecs/src/grfmt_pfm.cpp

```diff
@@ -0,0 +1,314 @@
+{
+  static const uint32_t A(0x000000ffU);
+  static const uint32_t B(0x0000ff00U);
+  static const uint32_t C(0x00ff0000U);
+  static const uint32_t D(0xff000000U);
+
+  union {
+    uint32_t i;
+    float f;
+  } cast;
+
+  cast.f = f;
+  cast.i = ( (cast.i & A) << 24 )
+         | ( (cast.i & B) <<  8 )
+         | ( (cast.i & C) >>  8 )
+         | ( (cast.i & D) >> 24 );
+  f = cast.f;
+}
+
+template<typename T> T atoT(const std::string& s);
+template<> int atoT<int>(const std::string& s) { return std::atoi(s.c_str()); }
+template<> float atoT<float>(const std::string& s) { return std::atof(s.c_str()); }
+
+template<typename T>
+T read_number(cv::RLByteStream& strm)
+{
+  // should be enogh to take string representation of any number
+  const size_t buffer_size = 2048;
+
+  std::vector<char> buffer(buffer_size, 0);
+  for (size_t i = 0; i < buffer_size; ++i) {
+    char c = strm.getByte();
+    if (std::isspace(c)) {
+      break;
+    }
+    buffer[i] = c;
+  }
+  const std::string str(buffer.begin(), buffer.end());
+  return atoT<T>(str);
+}
+
+template<typename T> void write_anything(cv::WLByteStream& strm, const T& t)
+{
+  const std::string str = std::to_string(t);
+  strm.putBytes(str.c_str(), str.size());
+}
+
+}
+
+namespace cv {
+
+PFMDecoder::~PFMDecoder()
+{
+}
+
+PFMDecoder::PFMDecoder()
+{
+  m_strm.close();
+}
+
+bool PFMDecoder::readHeader()
+{
+  if (m_buf.empty()) {
+    if (!m_strm.open(m_filename)) {
+      return false;
+    }
+  } else {
+    if (!m_strm.open(m_buf)) {
+      return false;
+    }
+  }
+
+  if (m_strm.getByte() != 'P') {
+    CV_Error(Error::StsError, "Unexpected file type (expected P)");
+    return false;
+  }
+
+  switch (m_strm.getByte()) {
+  case 'f':
+    m_channels = 1;
+    break;
+  case 'F':
+    m_channels = 3;
+    break;
+  default:
+    CV_Error(Error::StsError, "Unexpected file type (expected `f` or `F`)");
+    return false;
+  }
+
+  if ('\n' != m_strm.getByte()) {
+    CV_Error(Error::StsError, "Unexpected header format (expected line break)");
+    return false;
+  }
+
+
+  m_width = read_number<int>(m_strm);
+  m_height = read_number<int>(m_strm);
+  m_scale_factor = read_number<float>(m_strm);
+  m_swap_byte_order = is_byte_order_swapped(m_scale_factor);
+
+  return true;
+}
+
+bool PFMDecoder::readData(Mat& mat)
+{
+  if (!m_strm.isOpened()) {
+    CV_Error(Error::StsError, "Unexpected status in data stream");
+    return false;
+  }
+
+  if (m_channels == 3) {
+    mat = Mat3f(m_height, m_width);
+  } else if (m_channels == 1) {
+    mat = Mat1f(m_height, m_width);
+  } else {
+    CV_Error(Error::StsError, "Expected 1 or 3 channels.");
+    return false;
+  }
+
+  CV_Assert(abs(m_scale_factor) == 1.0);
+
+  for (int y = m_height - 1; y >= 0; --y) {
+    m_strm.getBytes(mat.ptr(y), m_width * mat.elemSize());
+    if (is_byte_order_swapped(m_scale_factor)) {
+      for (size_t i = 0; i < m_width * m_channels; ++i) {
+        swap_endianess(mat.ptr<float>(y)[i]);
```

Comment:
Perhaps it is better to swap `uint32_t` instead of `float`.

Reply:
`swap_endianess` already works on `uint32_t` internally. But it can be simplified, though.